File: DragGesture-OC/DragGesture-OC/TestVideoView.m
//
//  TestVideoView.m
//  DragGesture-OC
//
//  Created by 杜奎 on 2019/4/28.
//  Copyright © 2019 du. All rights reserved.
//

#import "TestVideoView.h"
#import <AVKit/AVKit.h>
#import "DragGestureHandler.h"

@interface TestVideoView ()

@property (nonatomic, strong) DragGestureHandler *gesture;
@property (nonatomic, strong) UIImageView *playBgView;
@property (nonatomic, strong) UIView      *playView;

@property (nonatomic, strong) AVPlayer *player;
@property (nonatomic, strong) AVPlayerLayer *playLayer;

@end

@implementation TestVideoView

- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        [self createUI];
    }
    return self;
}

- (void)createUI {
    [self addSubview:self.playBgView];
    [self addSubview:self.playView];
    [self addSubview:self.videoCoverView];
    
    __weak typeof(self) weakSelf = self;
    self.gesture = [[DragGestureHandler alloc] initWithGestureView:self.playView bgView:self.playBgView];
    self.gesture.completeBlock = ^(BOOL finish) {
        if (finish) {
            [weakSelf hide];
        }
    };
}

#pragma mark - action

- (void)show {
    [UIApplication.sharedApplication.keyWindow addSubview:self];
    self.playBgView.alpha = 0;
    self.playView.alpha = 0;
    
    self.videoCoverView.frame = self.outsideFrame;
    
    self.playView.frame = self.insideFrame;
    self.playLayer.frame = self.playView.bounds;
    [self.playView.layer addSublayer:self.playLayer];
    [self.player play];
    
    self.playBgView.userInteractionEnabled = NO;
    [UIView animateWithDuration:0.3 animations:^{
        self.playBgView.alpha = 1;
        self.videoCoverView.frame = self.insideFrame;
    } completion:^(BOOL finished) {
        self.playBgView.userInteractionEnabled = YES;
        self.playView.alpha = 1;
        self.videoCoverView.alpha = 0;
    }];
}

- (void)hide {
    self.playBgView.userInteractionEnabled = NO;
    [self.player pause];
    self.playView.alpha = 0;
    self.videoCoverView.alpha = 1;
    self.videoCoverView.frame = self.playView.frame;
    
    [UIView animateWithDuration:0.3 animations:^{
        self.playBgView.alpha = 0;
        self.videoCoverView.frame = self.outsideFrame;
    } completion:^(BOOL finished) {
        self.playBgView.userInteractionEnabled = YES;
        [self removeFromSuperview];
    }];
}

- (void)tagAction {
    [self hide];
}

#pragma mark - setter & getter

- (AVPlayer *)player {
    if (!_player) {
        NSString *path = [[NSBundle mainBundle] pathForResource:@"testVideo" ofType:@".MOV"];
        NSString *filePath = [NSString stringWithFormat:@"file://%@", path];
        NSURL *url = [[NSURL alloc] initWithString:filePath];
        AVPlayerItem *item = [[AVPlayerItem alloc] initWithURL:url];
        _player = [[AVPlayer alloc] initWithPlayerItem:item];
    }
    return _player;
}

- (AVPlayerLayer *)playLayer {
    if (!_playLayer) {
        _playLayer = [AVPlayerLayer playerLayerWithPlayer:self.player];
        _playLayer.videoGravity = AVLayerVideoGravityResize;
    }
    return _playLayer;
}

- (UIImageView *)playBgView {
    if (!_playBgView) {
        _playBgView = [[UIImageView alloc] initWithFrame:self.bounds];
        _playBgView.backgroundColor = UIColor.blackColor;
        _playBgView.userInteractionEnabled = NO;
        [_playBgView addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tagAction)]];
    }
    return _playBgView;
}

- (UIView *)playView {
    if (!_playView) {
        _playView = [[UIView alloc] init];
        _playView.backgroundColor = UIColor.clearColor;
        _playView.userInteractionEnabled = NO;
    }
    return _playView;
}

- (UIImageView *)videoCoverView {
    if (!_videoCoverView) {
        _videoCoverView = [[UIImageView alloc] init];
        _videoCoverView.contentMode = UIViewContentModeScaleAspectFill;
        _videoCoverView.clipsToBounds = YES;
    }
    return _videoCoverView;
}

@end

File: DragGesture-Swift/DragGesture-Swift/TestVideoView.swift
//
//  TestVideoView.swift
//  DragGesture-Swift
//
//  Created by 杜奎 on 2019/4/27.
//  Copyright © 2019 du. All rights reserved.
//

import UIKit
import AVKit

//此处视频播放仅仅用于展示手势效果，实际使用不可取
class TestVideoView: UIView {

    var outsideFrame = CGRect.zero
    var insideFrame = CGRect.zero
    var gesture: DragGestureHandler?
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        self.createUI()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func createUI() {
        self.addSubview(self.playBgView)
        self.addSubview(self.playView)
        self.addSubview(self.videoCoverView)
        
        self.gesture = DragGestureHandler.init(gestureView: self.playView, bgView: self.playBgView)
        self.gesture?.completeBlock = { [weak self] (finish) in
            if finish {
                self?.hide()
            }
        }
    }
    
    //MARK:- action
    
    func show() {
        UIApplication.shared.keyWindow?.addSubview(self)
        self.playBgView.alpha = 0
        self.playView.alpha = 0

        self.videoCoverView.frame = self.outsideFrame
        
        self.playView.frame = self.insideFrame
        self.playerLayer.frame = self.playView.bounds
        self.playView.layer.addSublayer(self.playerLayer)
        self.player.play()
        
        self.playBgView.isUserInteractionEnabled = false

        UIView.animate(withDuration: 0.3, animations: {
            self.playBgView.alpha = 1
            self.videoCoverView.frame = self.insideFrame
        }) { (finish) in
            self.playBgView.isUserInteractionEnabled = true
            self.playView.alpha = 1
            self.videoCoverView.alpha = 0
        }
    }
    
    func hide() {
        self.playBgView.isUserInteractionEnabled = false
        self.player.pause()
        self.playView.alpha = 0
        self.videoCoverView.alpha = 1
        self.videoCoverView.frame = self.playView.frame

        UIView.animate(withDuration: 0.3, animations: {
            self.playBgView.alpha = 0
            self.videoCoverView.frame = self.outsideFrame
        }) { (finish) in
            self.playBgView.isUserInteractionEnabled = true
            self.removeFromSuperview()
        }
    }
    
    @objc func tapAction() {
        self.hide()
    }
    
    //MARK:- setter & getter
    
    private lazy var player: AVPlayer = {
        let path = Bundle.main.path(forResource: "testVideo", ofType: ".MOV") ?? ""
        let filePath = "file://" + path
        let url = URL.init(string: filePath)
        let item = AVPlayerItem.init(url: url!)
        let player = AVPlayer.init(playerItem: item)
        return player
    }()
    
    private lazy var playerLayer: AVPlayerLayer = {
        let layer = AVPlayerLayer.init(player: self.player)
        layer.videoGravity = AVLayerVideoGravity.resize
        return layer
    }()

    private lazy var playBgView: UIView = {
        let view = UIView.init(frame: self.bounds)
        view.backgroundColor = UIColor.black
        view.isUserInteractionEnabled = false
        view.addGestureRecognizer(UITapGestureRecognizer.init(target: self, action: #selector(tapAction)))
        return view
    }()
    
    private lazy var playView: UIView = {
        let view = UIView.init()
        view.backgroundColor = UIColor.clear
        view.isUserInteractionEnabled = false
        return view
    }()
    
    private(set) var videoCoverView: UIImageView = {
        let view = UIImageView.init()
        view.contentMode = .scaleAspectFill
        view.clipsToBounds = true
        return view
    }()

}
